### PR TITLE
Add support for JSON/YAML serialization (and prepare stuff for deserialization)

### PR DESCRIPTION
**Pull request by haecker-felix**

### CODE REVIEW

File: src/collection.rs [hunk 1]
```diff
@@ -12,13 +12,24 @@ use std::io::prelude::*;
...
+#[serde(rename_all(serialize = "PascalCase"))]
 pub struct Collection {
+    #[serde(alias = "Version")]
     pub version: String,
-    #[serde(default)]
+    #[serde(default, skip_serializing_if = "Option::is_none", alias = "Origin")]
     pub origin: Option<String>,
-    #[serde(rename = "component", default)]
+    #[serde(
+        rename(deserialize = "component", serialize = "Component"),
```
Comment (bilelmoussaoui) [inline src/collection.rs:22]: Do we really need to specify the serialize part?

Comment (haecker-felix) [reply src/collection.rs:22]: Yes. In XML it's `component`, and for JSON it's `Component`. Otherwise the upper/lower case would not be correct.

Comment (bilelmoussaoui) [reply src/collection.rs:22]: Right but alias should fix it already

Comment (haecker-felix) [reply src/collection.rs]: Yup you're right, my bad. Fixed it.

Comment (haecker-felix) [reply src/collection.rs]: Aaah nope. It is required for serializing :(
Serializing doesn't care about the alias, so it takes the lowercase name.

File: src/component.rs [hunk 2]
```diff
@@ -23,70 +23,135 @@ pub struct Component {
     pub kind: ComponentKind,
     #[serde(deserialize_with = "app_id_deserialize")]
     pub id: AppId,
-    #[serde(rename = "name", deserialize_with = "translatable_deserialize")]
+    #[serde(deserialize_with = "translatable_deserialize")]
     pub name: TranslatableString,
+    #[serde(deserialize_with = "some_translatable_deserialize", default)]
+    #[serde(skip_serializing_if = "Option::is_none")]
```
Comment (bilelmoussaoui) [inline src/component.rs:29]: Nitpick: fusion both

File: src/enums.rs [hunk 3]
```diff
@@ -399,6 +399,64 @@ pub enum Icon {
...
+            pub name: Option<String>,
+            #[serde(skip_serializing_if = "Option::is_none")]
+            pub path: Option<PathBuf>,
+            #[serde(skip_serializing_if = "Option::is_none")]
+            pub url: Option<Url>,
+            #[serde(skip_serializing_if = "Option::is_none")]
+            pub width: Option<u32>,
+            #[serde(skip_serializing_if = "Option::is_none")]
+            pub height: Option<u32>,
+        }
```
Comment (bilelmoussaoui) [inline src/enums.rs:421]: Do we really need this? It was just a hackish thing I did for deserialize just to get the first release rolling; I will remove them from the deserialize part as well

Comment (haecker-felix) [reply src/enums.rs:421]: Feel free to make a suggestion how to improve this. IMO it's more elegant as doing `s.serialize_field("height", &height)?` manually for every field.

Comment (bilelmoussaoui) [reply src/enums.rs:421]: The thing is that for other structs/enums, you have done it the other so it makes things looks inconsistent :/

File: src/enums.rs [hunk 4]
```diff
@@ -380,7 +484,39 @@ pub enum Launchable {
...
 
-#[derive(Clone, Debug, Serialize, PartialEq)]
+impl Serialize for Launchable {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut s = serializer.serialize_struct("launchable", 2)?;
+        match self {
+            Launchable::DesktopId(url) => {
```
Comment (bilelmoussaoui) [inline src/enums.rs:494]:
```suggestion
            Launchable::DesktopId(app_id) => {
```

File: src/enums.rs [hunk 5]
```diff
@@ -380,7 +484,39 @@ pub enum Launchable {
...
+                s.serialize_field("type", "desktop_id")?;
+                s.serialize_field("name", &url)?;
+            }
+            Launchable::Service(url) => {
```
Comment (bilelmoussaoui) [inline src/enums.rs:498]:
```suggestion
            Launchable::Service(name) => {
```

File: src/enums.rs [hunk 6]
```diff
@@ -380,7 +484,39 @@ pub enum Launchable {
...
+                s.serialize_field("type", "service")?;
+                s.serialize_field("name", &url)?;
+            }
+            Launchable::Url(url) => {
+                s.serialize_field("type", "url")?;
+                s.serialize_field("name", &url)?;
+            }
+            Launchable::CockpitManifest(url) => {
```
Comment (bilelmoussaoui) [inline src/enums.rs:506]:
```suggestion
            Launchable::CockpitManifest(manifest) => {
```

File: src/enums.rs [hunk 7]
```diff
@@ -380,7 +484,39 @@ pub enum Launchable {
...
+                s.serialize_field("type", "cockpit_manifest")?;
+                s.serialize_field("name", &url)?;
+            }
+            Launchable::Unknown(url) => {
```
Comment (bilelmoussaoui) [inline src/enums.rs:510]: I would say, just ignore that url?

File: src/enums.rs [hunk 8]
```diff
@@ -392,6 +528,50 @@ pub enum ProjectUrl {
...
+impl Serialize for ProjectUrl {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut s = serializer.serialize_struct("url", 2)?;
+        match self {
+            ProjectUrl::Donation(url) => {
+                s.serialize_field("type", "donation")?;
+                s.serialize_field("url", &url)?;
```
Comment (bilelmoussaoui) [inline src/enums.rs:540]: You can do s.serialize_filed("url", &self.0)?; only once for all the project urls

Comment (haecker-felix) [reply src/enums.rs:540]:
```
error[E0609]: no field `0` on type `&enums::ProjectUrl`
   --> src/enums.rs:571:40
    |
571 |         s.serialize_field("url", &self.0)?;
    |                                        ^

```

File: src/enums.rs (hunk 7 above)
Comment (haecker-felix) [reply src/enums.rs]: I don't think that would be a good idea. We should serialize the whole `Component` struct completely and don't ignore stuff.

**Pull request merged**